### PR TITLE
NEW Add merge-ups

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/gha-merge-up/issues/4

Also updates the logic for license

### CODE REVIEW

**Comment by GuySartorelli** [comment]
Please add information about what these changes actually are - they're obviously going beyond what the purpose of the issue is. You mentioned something wasn't working as an off-hand comment in person so I assume this is mostly bug fixes? What were the bugs? What are you doing here and why?

**Comment by emteknetnz** [comment]
I experienced loads of issues when running this for real. Some things I do remember going wrong:
- Realising after creating heaps of PRs that we need keepalive.yml in any repo that has merge-up.yml in it
- Creating the option "--update-prs" so that existing PRs were updated rather than having to close all the current open oes
- Issues when there was a newer version of the automatically created branches and using "--update-prs"
- Adding `rename_file_if_exists()` because there was some repo that had 2x license files in it
- Needing to add gha-* modules
- Outputting PHP errors to console while still showing which PRs were created
- Special handling for weird repos - developer-docs, silverstripe-simple, cwp-watea-theme

File: scripts/cms5/keepalive.php
Comment (GuySartorelli) [inline scripts/cms5/keepalive.php:4]:
```suggestion
// run on a day of the month up to the 28th
$runOnDay = predictable_random_int(28);
```
And make sure it can't be 0

File: funcs_scripts.php
Comment (GuySartorelli) [inline funcs_scripts.php:213]:
```suggestion
 * Creates a predicatable random int between 0 and $max based on the module name to be used with the % mod operator.
 * Note that this will return the exact same value every time it is called for a given module.
```
Just to make this abundantly clear.

File: funcs_scripts.php
Comment (GuySartorelli) [inline funcs_scripts.php:220]:
```suggestion
    mt_srand(array_sum($codes));
    return mt_rand(0, $max);
```
May as well use build in random int functionality. Has the benefit of being clearer that it's producing a (predictably) random int within the limits without having to mentally walk through the math.

File: scripts/cms5/merge-ups.php
Comment (GuySartorelli) [inline scripts/cms5/merge-ups.php:8]: These comments are misplaced, since the 3 days offset and other "same as CI run" things happen later.
```suggestion
// run on a random day of the week
$runOnDay = predictable_random_int(6);
// run at a random hour of the day
$runOnHour = predictable_random_int(23);
// run at a random minute of the hour rounded to 5 minutes
```

File: scripts/cms5/merge-ups.php
Comment (GuySartorelli) [inline scripts/cms5/merge-ups.php:11]:
```suggestion
// If there's a CI workflow, offset mergeups from the CI run by 3 days
if (check_file_exists('.github/workflows/dispatch-ci.yml')) {
```

File: scripts/cms5/merge-ups.php
Comment (GuySartorelli) [inline scripts/cms5/merge-ups.php:25]: Don't think it's necessary to run at the same hour and minute - but I don't feel strongly enough to explicitly ask for it to be removed if you think it makes sense to have there.

File: update_command.php
Comment (GuySartorelli) [inline update_command.php:7]: What's this here for? I can't find it being used?

File: update_command.php
Comment (GuySartorelli) [inline update_command.php:44]: Please add a comment in code about why this is only for current major.
I was going to just ask you to explain it to me directly but a code comment seems better for this as a) I'm not the only person likely to be confused and b) I'll be confused about it next time I see it anyway lol

File: update_command.php
Comment (GuySartorelli) [inline update_command.php:137]: This seems sensible - except that this means any changes which _need_ to be on the default branch to work (e.g. changes to cron-based workflows) simply won't do anything. Do you have any thoughts about those scenarios? Fluent for example has `master` as its default branch.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Once this PR has been merged, it'll need to be run again since the number generator implementation has changed. Best to wait until its merged so we don't have auto-generated PRs being merged and then recreated and merged and etc etc if this needs multiple rounds of review.

## About the new "update-prs" option

This option is confusing to me. I would expect that this would update PRs if they exist, but _still run on all modules whether there is a PR or not_.
e.g. I run it once, and it creates some PRs. I then run it again, before the PRs have been merged, and it updates PRs that exist and also runs against any modules that were missed the first time around.

The above seems like a far more likely scenario than wanting to _only_ run the scripts against repos which already have PRs. What is the intended scenario for the current implementation, where it will skip repositories which don't have PRs?

Note that I haven't reviewed the code changes in update_command.php related to that new option pending a response to this question as I really don't understand the intended use case.

File: funcs_scripts.php
Comment (emteknetnz) [reply funcs_scripts.php:133]: vendor/plugin is a weird one. I wouldn't say it's used by the cms, instead I'd say it's used by composer

It's a requirement of silverstripe/installer so that can do things like run `composer vendor-expose`

File: funcs_scripts.php
Comment (emteknetnz) [reply funcs_scripts.php:220]: It's not meant to be an actually random number, it's means to be "seemingly random though actually gives the exact same result each time"

File: scripts/cms-any/editorconfig.php
Comment (emteknetnz) [reply scripts/cms-any/editorconfig.php:39]: the .editorconfig is kind of rough e.g. includes `[admin/thirdparty/**]`

Seems like should have something nicer for the non-cms repos like eslint-config, though writing a new editor config is getting way outta scope for this card

File: update_command.php
Comment (emteknetnz) [reply update_command.php:137]: Our standard is integer's for next-minors. If fluent doesn't play nice that's because fluent is non-standard and the solution is to standardise fluent, not to loosen things here

**Comment by emteknetnz** [comment]
> I would expect that this would update PRs if they exist, but still run on all modules whether there is a PR or not.

Yeah that seems like a really nice use of this new flag.  Previously intention was to use copy-paste a list of repos run into the --exclude flag to things broke half-way through, though you could simply use --update-prs instead which seems like it would be much nicer.

Though could we split this as a separate card. This PR is already very convoluted and implementing this new would further convoluted it

File: funcs_scripts.php
Comment (emteknetnz) [reply funcs_scripts.php:85]: Updated

File: funcs_scripts.php
Comment (emteknetnz) [reply funcs_scripts.php:213]: Updated

File: funcs_utils.php
Comment (emteknetnz) [reply funcs_utils.php:76]: Updated

File: funcs_utils.php
Comment (emteknetnz) [reply funcs_utils.php:80]: Updated, it's very easy

File: funcs_utils.php
Comment (emteknetnz) [reply funcs_utils.php:88]: Updated

File: funcs_utils.php
Comment (emteknetnz) [reply funcs_utils.php:200]: Updated

File: funcs_utils.php
Comment (emteknetnz) [reply funcs_utils.php:302]: Leaving as is

File: run.php
Comment (emteknetnz) [reply run.php:86]: Updated

File: scripts/cms5/keepalive.php
Comment (emteknetnz) [reply scripts/cms5/keepalive.php:4]: Updated

File: scripts/cms5/merge-ups.php
Comment (emteknetnz) [reply scripts/cms5/merge-ups.php:8]: Updated

File: scripts/cms5/merge-ups.php
Comment (emteknetnz) [reply scripts/cms5/merge-ups.php:11]: Updated

File: scripts/cms5/merge-ups.php
Comment (emteknetnz) [reply scripts/cms5/merge-ups.php:25]: Leave as is

File: update_command.php
Comment (emteknetnz) [reply update_command.php:7]: Typo, removed

File: update_command.php
Comment (emteknetnz) [reply update_command.php:44]: Updated.  I also renamed it from extra_modules() to extra_repositories() since gha-* aren't modules

File: funcs_scripts.php
Comment (GuySartorelli) [inline funcs_scripts.php:125]:
```suggestion
 * Determine if the repository represents a silverstripe module.
 * For the purposes of this module, recipes and themes are modules.
 *
 * Example usage:
 * is_module()
 */
function is_module()
```
You'll need to adjust the logic accordingly. Probably best would be to check if there's a composer.json file. If no composer, it's not a module. Then you can check the `type` key in composer.json for silverstripe-vendormodule, silverstripe-module (can skip this one if you like, it was the ss3 type), silverstripe-theme, and silverstripe-recipe

You could call it `is_not_module()` but then you end up with `if (!is_not_module())`, and double negatives are bad.

File: scripts/cms-any/editorconfig.php
Comment (GuySartorelli) [reply scripts/cms-any/editorconfig.php:39]: That thirdparty glob should probably be `**/thirdparty/**` but as you say it's out of scope for this card.
Arguably 90% of this PR should have been on a separate card that blocked this one, but that's sort of irrelevant to whether _this_ should be included here.

There's a new recommendation to make the use case of `module_is_not_for_cms()` much clearer which I think resolves my problems with this anyway.

File: funcs_scripts.php
Comment (GuySartorelli) [reply funcs_scripts.php]: As discussed in person, seeding with `mt_srand()` and then running `mt_rand()` will always give you the same random value every time - and it has the stated advantage of being clearly what's happening and therefore we don't need a unit test for it.

File: update_command.php
Comment (GuySartorelli) [reply update_command.php:137]: I'm on the fence about this, since I don't think it's likely we'll ever get the maintainer to follow our branching strategy. But that's a much wider discussion than should be held on this PR so I'm okay with leaving this as is for now.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Remaining changes (I forgot to hit start review):
- https://github.com/silverstripe/module-standardiser/pull/2#discussion_r1292897888
- https://github.com/silverstripe/module-standardiser/pull/2#discussion_r1292932330

File: funcs_scripts.php
Comment (emteknetnz) [reply funcs_scripts.php:125]: Updated

File: funcs_scripts.php
Comment (emteknetnz) [reply funcs_scripts.php]: Updated

File: funcs_scripts.php
Comment (GuySartorelli) [reply funcs_scripts.php]: Please create a card to update logic as per above

Comment (emteknetnz) [reply funcs_scripts.php]: https://github.com/silverstripe/module-standardiser/issues/4

**Review by GuySartorelli — APPROVED** [review]
Looks okay. There are some follow-up cards created for things that weren't quite optimal but can be dealt with later.

**Pull request merged**
